// dimensions-ai/lib/Design/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var __1 = require("..");
/**
 * @class Design
 * @classdesc Abstract class detailing a `Design` to be used as the platform that holds competition runtime logic for
 * updating and manipulating ongoing `matches`
 *
 * Refer to `Match` class for exposed fields available for user's use. User's can also extend the `Match` class and add
 * more fields if they wish for their own use and pass their own `Match` class instantiations
 *
 */
var Design = /** @class */ (function () {
    function Design(name, designOptions) {
        this.name = name;
        this.log = new __1.Logger();
        // TODO: [Strange] - This design option breaks jest when placed outside, jest thinks LoggerLEVEL.info isn't real
        var DEFAULT_DESIGN_OPTIONS = {
            commandStreamType: COMMAND_STREAM_TYPE.SEQUENTIAL,
            commandDelimiter: ','
        };
        // Set defaults
        this.designOptions = DEFAULT_DESIGN_OPTIONS;
        // Override with user provided params
        Object.assign(this.designOptions, designOptions);
        // Set log level to default
        this.log.level = __1.Logger.LEVEL.INFO;
        this.log.system("Initialized Design: " + this.name);
    }
    Design.prototype._setLogLevel = function (level) {
        this.log.level = level;
    };
    /**
     * Get the design options associated with this `Design`
     */
    Design.prototype.getDesignOptions = function () {
        return this.designOptions;
    };
    return Design;
}());
exports.Design = Design;
// Standard ways for commands from agents to be streamed to `MatchEngine` for the `Design` to handle
var COMMAND_STREAM_TYPE;
(function (COMMAND_STREAM_TYPE) {
    COMMAND_STREAM_TYPE[COMMAND_STREAM_TYPE["PARALLEL"] = 0] = "PARALLEL";
    COMMAND_STREAM_TYPE[COMMAND_STREAM_TYPE["SEQUENTIAL"] = 1] = "SEQUENTIAL"; // each agent's set of command sequence is run before the next agent
})(COMMAND_STREAM_TYPE = exports.COMMAND_STREAM_TYPE || (exports.COMMAND_STREAM_TYPE = {}));
;
